### PR TITLE
add files for code and learn at COSCon 2019

### PR DESCRIPTION
**Pull request by joyeecheung**

### CODE REVIEW

**Comment by joyeecheung** [comment]
cc @MoonBall @oyyd @ZYSzys @starkwang @legendecas

File: archive/shanghai-2019/README.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Node.js Code + Learn @ COSCon 2019
+
+- [事前准备(praparation)](./preparation.md)
+- [Dockerfile](./Dockerfile)
```
Comment (oyyd) [inline archive/shanghai-2019/README.md:4]: Also add the [introduction presentation](https://docs.google.com/presentation/d/1NluODLOelMFyui10jSLY8T4M0F444LaG7Ir5u_tfOGM/edit#slide=id.p1) here?

File: archive/shanghai-2019/Dockerfile [hunk 2]
```diff
@@ -21,8 +21,8 @@ WORKDIR /home/node-dev
 ENV HOME /home/node-dev
 
 # Clone the code and build
-git clone --depth=1 https://github.com/nodejs/node.git
+RUN git clone --depth=1 https://github.com/nodejs/node.git
 WORKDIR /home/node-dev/node
 RUN python2 ./configure --ninja
```
Comment (cclauss) [inline archive/shanghai-2019/Dockerfile:26]: Why hardcode legacy Python?  Node.js currently prefers `python3` over `python2`.

Comment (legendecas) [reply archive/shanghai-2019/Dockerfile:26]: I've just failed with python3 yesterday... didn't look into the error.

Comment (joyeecheung) [reply archive/shanghai-2019/Dockerfile:26]: Because the master does not currently work with python3 and we need to have a running Dockerfile by today.

Comment (cclauss) [reply archive/shanghai-2019/Dockerfile:26]: This Dockerfile failed??

Comment (legendecas) [reply archive/shanghai-2019/Dockerfile:26]: Running with python 3.7.4:

```
⋊> ~/D/g/node on master ◦ python3 ./configure.py -C
INFO: Using floating patch "tools/icu/patches/64/source/common/putil.cpp" from "tools/icu"
INFO: Using floating patch "tools/icu/patches/64/source/i18n/dtptngen.cpp" from "tools/icu"
No receipt for 'com.apple.pkg.CLTools_Executables' found at '/'.
No receipt for 'com.apple.pkg.DeveloperToolsCLILeo' found at '/'.
No receipt for 'com.apple.pkg.DeveloperToolsCLI' found at '/'.
gyp: No Xcode or CLT version detected!
Error running GYP
⋊> ~/D/g/node on master ◦ python ./configure.py -C
INFO: Using floating patch "tools/icu/patches/64/source/common/putil.cpp" from "tools/icu"
INFO: Using floating patch "tools/icu/patches/64/source/i18n/dtptngen.cpp" from "tools/icu"
INFO: configure completed successfully
```

Opened issue https://github.com/nodejs/node/issues/30215.

Comment (joyeecheung) [reply archive/shanghai-2019/Dockerfile:26]: BTW this is what currently happens if this command is switched to python3

```
Traceback (most recent call last):
  File "./configure", line 20, in <module>
    from distutils.spawn import find_executable
ModuleNotFoundError: No module named 'distutils.spawn'
```

Comment (richardlau) [reply archive/shanghai-2019/Dockerfile:26]: > 
> 
> BTW this is what currently happens if this command is switched to python3
> 
> ```
> Traceback (most recent call last):
>   File "./configure", line 20, in <module>
>     from distutils.spawn import find_executable
> ModuleNotFoundError: No module named 'distutils.spawn'
> ```

See https://github.com/nodejs/node/issues/30189.

File: archive/shanghai-2019/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,28 @@
+FROM ubuntu:bionic
+
+# Install tools
+RUN apt-get update
+RUN apt-get install -y build-essential python ninja-build make g++ vim git sudo
```
Comment (legendecas) [inline archive/shanghai-2019/Dockerfile]: `make` and `g++` are optional, which are dependencies of `build-essential`.

File: archive/shanghai-2019/Dockerfile (hunk 2 above)
Comment (joyeecheung) [reply archive/shanghai-2019/Dockerfile:26]: It still fails with 

```
Traceback (most recent call last):
  File "../../deps/v8/tools/gen-postmortem-metadata.py", line 693, in <module>
    load_objects();
  File "../../deps/v8/tools/gen-postmortem-metadata.py", line 375, in load_objects
    load_objects_from_file(filename, checktypes)
  File "../../deps/v8/tools/gen-postmortem-metadata.py", line 397, in load_objects_from_file
    for line in objfile:
  File "/usr/lib/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 2624: ordinal not in range(128)
```

after installing `python3-distutils`.

We need to give the organizers this Dockerfile ASAP. Is there a way to get this fixed in 1 hour?

**Comment by joyeecheung** [comment]
Can we merge this ASAP? This is an archive to keep track of the things that we give to the organizers and the participants, it's not supposed to be a perfect example.

**Review by oyyd — APPROVED** [review]

File: archive/shanghai-2019/Dockerfile (hunk 2 above)
Comment (cclauss) [reply archive/shanghai-2019/Dockerfile:26]: Please stick with `python2` for now.

Comment (cclauss) [reply archive/shanghai-2019/Dockerfile:26]: @legendecas Thank you very much!  This Dockerfile has allowed me to find and propose fixes for nodejs/node#30218 and nodejs/node#30220.

**Pull request merged**